kv/txn.go

```diff
@@ -19,21 +19,124 @@ import (
 	"errors"
 	"math"
 	"math/rand"
+	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/parser/terror"
 	"github.com/pingcap/tidb/util/logutil"
+	"github.com/tikv/client-go/v2/oracle"
 	"go.uber.org/zap"
 )
 
+const (
+	chanBufferSize             = 256
+	undeletedStartTsBufferSize = 16
+	// TimeToPrintLongTimeInternalTxn is the duration if the internal transaction lasts more than it, TiDB prints
+	// a log message
+	TimeToPrintLongTimeInternalTxn = time.Minute
+)
+
+var globalInnerTxnTsBox atomic.Value
+
+type innerTxnStartTsBox struct {
+	innerTSLock        sync.Mutex
+	innerTxnStartTsMap map[uint64]uint64
+}
+
+// InitInnerTxnStartTsBox initializes globalInnerTxnTsBox
+func InitInnerTxnStartTsBox() {
+	iTxnTsBox := &innerTxnStartTsBox{
+		innerTxnStartTsMap: make(map[uint64]uint64, chanBufferSize),
+	}
+	globalInnerTxnTsBox.Store(iTxnTsBox)
+}
+
+func getGlobalInnerTxnTsBox() *innerTxnStartTsBox {
+	v := globalInnerTxnTsBox.Load()
+	if v == nil {
+		return nil
+	}
+	ib := v.(*innerTxnStartTsBox)
+	return ib
+}
+
+// wrapStoreInterTxnTS is the entry function used to store the startTS of an internal transaction to globalInnerTxnTsBox
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapStoreInterTxnTS(startTS uint64) {
+	ib := getGlobalInnerTxnTsBox()
+	if ib == nil {
+		return
+	}
+	ib.storeInnerTxnTS(startTS)
+}
+
+func (ib *innerTxnStartTsBox) storeInnerTxnTS(startTS uint64) {
+	ib.innerTSLock.Lock()
+	ib.innerTxnStartTsMap[startTS] = startTS
+	ib.innerTSLock.Unlock()
+
+}
+
+// wrapDeleteInterTxnTS delete the startTS from globalInnerTxnTsBox when the transaciotn is commted
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapDeleteInterTxnTS(startTS uint64) {
+	ib := getGlobalInnerTxnTsBox()
+	if ib == nil {
+		return
+	}
+	ib.deleteInnerTxnTS(startTS)
+}
+
+func (ib *innerTxnStartTsBox) deleteInnerTxnTS(startTS uint64) {
+	ib.innerTSLock.Lock()
+	delete(ib.innerTxnStartTsMap, startTS)
+	ib.innerTSLock.Unlock()
+}
+
+// WrapGetMinStartTs get the min StatTS between startTSLowerLimit and curMinStartTS in globalInnerTxnTsBox
+func WrapGetMinStartTs(now time.Time, startTSLowerLimit uint64,
+	curMinStartTS uint64) uint64 {
+	ib := getGlobalInnerTxnTsBox()
+	if ib == nil {
+		logutil.BgLogger().Info("Fail to get internal txn startTS, globalInnerTxnTsBox is nil")
+		return curMinStartTS
+	}
+	return ib.getMinStartTs(now, startTSLowerLimit, curMinStartTS)
+}
+
+func (ib *innerTxnStartTsBox) getMinStartTs(now time.Time, startTSLowerLimit uint64,
+	curMinStartTS uint64) uint64 {
```

Comment:
IMHO `curMinStartTS` needn't be passed into this function, so the function itself does exactly one thing, instead of mixing getting min startTS and combining it with other timestamps.

Reply:
If do that, the caller must compare the returned startTS with current min startTS and get the smaller. It looks not so cleanly. And I think the function getMinStartTs doesn't mix more things together, It just gets the minimum startTS which consider the current one.

Reply:
Well, I prefer to compare current min start TS out of the function because in my mind the logic is (1) get min start TS among internal sessions (2) get min startTS among `RunInNewTxn`s (3) use the smaller one
Each step is independent. It's not necessary for step (2) to depend on the min start TS got from step (1).

Reply:
Yes, the logic looks clear. But the original code mixed minStartTS and startTSLowerLimit with getting the minimum startTS for client sessions as bellow. 
	for _, info := range pl {
		if info.CurTxnStartTS > startTSLowerLimit && info.CurTxnStartTS < minStartTS {
			minStartTS = info.CurTxnStartTS
		}
	} 
In actually , I intended to wrap this code into a function too because of similar logic. But the structures for store TS are different and I give it up.
And I think it can remove minStartTS from the function getMinStartTs, but can't remove startTSLowerLimit . So, I put it together.